docassemble/ALToolbox/test_al_income.py

```diff
@@ -0,0 +1,153 @@
+        self.assertEqual(Decimal("10.1"), income.total(times_per_year=12))
+
+        hourly_income = ALIncome(
+            value=4.4, times_per_year=52, is_hourly=True, hours_per_period=39
+        )
+        self.assertEqual(Decimal("8923.2"), hourly_income.total())
+        self.assertEqual(Decimal("171.6"), hourly_income.total(52))
+
+    def test_income_list(self):
+        income = ALIncome(source="coding", value=12.53, times_per_year=12)
+        hourly_income = ALIncome(
+            source="coding",
+            value=4.4,
+            times_per_year=2,
+            is_hourly=True,
+            hours_per_period=41,
+        )
+        income_list = ALIncomeList(elements=[income, hourly_income])
+        self.assertSetEqual(set(["coding"]), income_list.sources())
+
+        self.assertEqual(Decimal(0), income_list.total(0))
+        self.assertEqual(Decimal("511.16"), income_list.total(1))
+        self.assertEqual(Decimal("511.16"), income_list.total(1, source="coding"))
+        self.assertEqual(Decimal(0), income_list.total(1, source="wrong job"))
+        self.assertEqual(
+            Decimal("511.16"), income_list.total(1, source=["coding", "wrong job"])
+        )
+
+    def test_job(self):
+        # TODO
+        pass
+
+    def test_job_list(self):
+        # TODO
+        pass
+
+    def test_asset(self):
+        home = ALAsset(market_value=1234567.89, source="home")
+        self.assertEqual(Decimal(0), home.total())
+        savings = ALAsset(
+            balance=12.34, value=0.12, times_per_year=12, source="savings account"
+        )
+        self.assertEqual(Decimal("1.44"), savings.total())
+
+    def test_asset_list(self):
+        home = ALAsset(market_value=1234567.89, source="home")
+        savings = ALAsset(
+            balance=12.34, value=0.12, times_per_year=12, source="savings"
+        )
+        investing = ALAsset(
+            balance=23.45, value=1.2, times_per_year=12, source="stocks"
+        )
+        asset_list = ALAssetList(elements=[home, savings, investing])
+        self.assertEqual(Decimal("1234567.89"), asset_list.market_value(source="home"))
+        self.assertEqual(
+            Decimal("35.79"), asset_list.balance(source=["savings", "stocks"])
+        )
+        self.assertEqual(Decimal("15.84"), asset_list.total())
+
+    def test_vehicle(self):
+        # TODO
+        pass
+
+    def test_vehicle_list(self):
+        # TODO
+        pass
+
+    def test_itemized_value_dict(self):
+        itemized_dict = ALItemizedValueDict(
+            elements={
+                "val1": ALItemizedValue(
+                    value=5.30, exists=True, is_hourly=False, times_per_year=1
+                ),
+                "val2": ALItemizedValue(
+                    value=5.30, exists=False, is_hourly=False, times_per_year=1
+                ),
+            }
+        )
+        itemized_dict.hook_after_gather()
+        self.assertEqual(1, len(itemized_dict))
```

Comment:
Not relevant to this review: Would love to talk about this sometime and understand a little more.

Reply:
The quick gist is that I wanted to test the behavior of `hook_after_gather`, which removes  elements where `x.exists == False`. IMO, I'm not sure that's the best behavior of the function, but it should be tested at least.